runtime/src/in_mem_accounts_index.rs

```diff
@@ -1302,4 +1350,185 @@ mod tests {
+        assert_eq!(reclaims, expected_reclaims);
+
+        // replace other and new_slot
+        let mut slot_list = vec![(unique_other_slot, other_value), (new_slot, other_value)];
+        let expected_reclaims = slot_list.clone();
+        let other_slot = Some(unique_other_slot);
+        // upserting into slot_list that already contain an entry at 'new-slot', so do NOT addref
+        let mut reclaims = Vec::default();
+        assert!(
+            !InMemAccountsIndex::update_slot_list(
+                &mut slot_list,
+                new_slot,
+                info,
+                other_slot,
+                &mut reclaims,
+                previous_slot_entry_was_cached
+            ),
+            "other_slot: {:?}",
+            other_slot
+        );
+        assert_eq!(slot_list, vec![at_new_slot]);
+        assert_eq!(
+            reclaims,
+            expected_reclaims.into_iter().rev().collect::<Vec<_>>()
+        );
+
+        // nothing will exist at this slot
+        let missing_other_slot = unique_other_slot + 1;
+        let ignored_slot = 10; // bigger than is used elsewhere in the test
+        let ignored_value = info + 10;
+
+        let mut possible_initial_slot_list_contents;
+        // build a list of possible contents in the slot_list prior to calling 'update_slot_list'
+        {
+            // up to 3 ignored slot account_info (ignored means not 'new_slot', not 'other_slot', but different slot #s which could exist in the slot_list initially)
+            possible_initial_slot_list_contents = (0..3)
+                .into_iter()
+                .map(|i| (ignored_slot + i, ignored_value + i))
+                .collect::<Vec<_>>();
+            // account_info that already exists in the slot_list AT 'new_slot'
+            possible_initial_slot_list_contents.push(at_new_slot);
+            // account_info that already exists in the slot_list AT 'other_slot'
+            possible_initial_slot_list_contents.push((unique_other_slot, other_value));
+        }
+
+        /*
+         * loop over all possible permutations of 'possible_initial_slot_list_contents'
+         * some examples:
+         * []
+         * [other]
+         * [other, new_slot]
+         * [new_slot, other]
+         * [dummy0, new_slot, dummy1, other] (and all permutation of this order)
+         * [other, dummy1, new_slot] (and all permutation of this order)
+         * ...
+         * [dummy0, new_slot, dummy1, other_slot, dummy2] (and all permutation of this order)
+         */
+        let mut attempts = 0;
+        // loop over each initial size of 'slot_list'
+        for initial_slot_list_len in 0..=possible_initial_slot_list_contents.len() {
+            // loop over every permutation of possible_initial_slot_list_contents within a list of len 'initial_slot_list_len'
+            for content_source_indexes in
+                (0..possible_initial_slot_list_contents.len()).permutations(initial_slot_list_len)
```

Comment:
Would [`.powerset()`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.powerset) work more like the examples you drew out above?

Reply:
looks like no.
it returns:
[1,2]
[1,3]
[2,3]
but not [2,1], [3,1], [3,2]
I could do a powerset and an inner permutations(). What I have seems simple enough. I wasn't sure what to call what I want. powerset_with_permutations is what I'm after. Maybe such a thing exists?

Reply:
another [fun, but not quite right one.](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.multi_cartesian_product)

Reply:
Ah gotcha. Sorry for sending you down this rabbit hole.